ports/esp32/README.ota.md

```diff
@@ -0,0 +1,405 @@
+﻿# Over the air updates
+
+This is an example of how to do ota-updates using micropython and a very simple bottle server application. It assumes you have some way of notifying your esp of updates and transmit the update hashe securely (for example using ssl with certificate pinning).
+
+## Folders
+
+We are using the following folder structure:
+```
+/
+├─ boot.py
+├─┬ factory
+│ ├── VERSION
+│ └── main.py  
+├─ ota_0
+└─ ota_1
+```
+
+The folders "factory", "ota_o" and "ota_1" musst have corresponding partitions. "VERSION" contains our currently running software version identifier, for example a git commit hash.
+
+## Basic concept
+Our boot.py checks for the currently running partition, changes the working directory to the corresponding folder and exits. Micropython then searches for a main.py in the current working directory and executes it.
+
+When we do an update we overwrite a (currently not running) partition with our new micropython version and download a tar archiv with our new python code. We extract the python code in the corresponding folder and set a new boot partition. Then we reboot.
+Optionaly we want to be able to only update the micropython interpreter or the python code. To accomplish that we copy the currently running version if we do not get one to update.
+
+## Example boot.py
+```python
+import uos
+import esp
+import sys
+
+def bytecompare(a,b):
```

Comment:
Why is this function necessary, isn't it enough to just do `a == b`?  Or do I miss something?

Reply:
Its not, (a == b) can be = false while bytecompare(a,b) returns true, because its compare by referenz and compare by value

Reply:
Can you give an example where it's not the same?  MicroPython compares by value, eg:
```python
>>> bytearray(b'\x01\x02\x03')==b'\x01\x02\x03'
True
>>> bytearray(b'\x01\x02\x03')==bytearray(b'\x01\x02\x03')
True
>>> b'\x01\x02\x03'==bytearray(b'\x01\x02\x03')
True
```